File: go-sdk/examples/dust_destroy.go
// Package examples provides reusable Ifx business transaction planners.
package examples

import (
	"github.com/gagliardetto/solana-go"
	"github.com/ifx-run/ifx/go-sdk/expr"
	"github.com/ifx-run/ifx/go-sdk/ifelse"
	"github.com/ifx-run/ifx/go-sdk/patch"
	"github.com/ifx-run/ifx/go-sdk/patchedcpi"
	"github.com/ifx-run/ifx/go-sdk/scratch"
	"github.com/ifx-run/ifx/go-sdk/spltoken"
)

// DustThresholdRaw is the raw balance cutoff (NOT UI amount).
const DustThresholdRaw uint64 = 1000

// DustDestroyAccounts wire accounts for Token-2022 dust destroy.
type DustDestroyAccounts struct {
	Mint            solana.PublicKey
	TokenAccount    solana.PublicKey
	Owner           solana.PublicKey
	OwnerSigner     bool
	RentDestination solana.PublicKey
}

// PlanDustDestroyInstructions returns reset + let + three if_else steps (mirrors planDustDestroyTx).
func PlanDustDestroyInstructions(s *scratch.FrameScratch, accts DustDestroyAccounts) ([]solana.Instruction, error) {
	mint := accts.Mint
	tokenAccount := accts.TokenAccount
	owner := accts.Owner
	rentDestination := accts.RentDestination

	out := []solana.Instruction{s.IxReset()}

	b := s.LetBuilder()
	amount, err := b.SplToken2022Amount(tokenAccount)
	if err != nil {
		return nil, err
	}
	withheld, err := b.SplToken2022TransferFeeWithheld(tokenAccount)
	if err != nil {
		return nil, err
	}
	decimals, err := b.SplToken2022MintDecimals(mint)
	if err != nil {
		return nil, err
	}
	letIx, err := b.BuildIx()
	if err != nil {
		return nil, err
	}
	out = append(out, letIx)

	dust := expr.Lt(expr.Ref(amount.Index), expr.U64(DustThresholdRaw))

	burnTpl := spltoken.BurnCheckedInstruction(tokenAccount, mint, owner, accts.OwnerSigner)
	burn, err := patchedcpi.RawCpi(
		burnTpl,
		patch.RawCpiPatch(1, amount),
		patch.RawCpiPatch(9, decimals),
	).Build(nil)
	if err != nil {
		return nil, err
	}
	burnArgs, err := ifelse.Args(
		expr.And(dust, expr.NonZero(expr.Ref(amount.Index))),
		ifelse.Cpi(burn.Cpi),
		ifelse.Skip,
	)
	if err != nil {
		return nil, err
	}
	burnIfElse, err := s.IxIfElse(burnArgs, burn.Remaining)
	if err != nil {
		return nil, err
	}
	out = append(out, burnIfElse)

	harvestTpl := spltoken.HarvestWithheldTokensToMintInstruction(mint, tokenAccount)
	harvestCpi, harvestRem, err := patchedcpi.StaticCpi(harvestTpl, nil)
	if err != nil {
		return nil, err
	}
	harvestArgs, err := ifelse.Args(
		expr.And(dust, expr.NonZero(expr.Ref(withheld.Index))),
		ifelse.Cpi(harvestCpi),
		ifelse.Skip,
	)
	if err != nil {
		return nil, err
	}
	harvestIfElse, err := s.IxIfElse(harvestArgs, harvestRem)
	if err != nil {
		return nil, err
	}
	out = append(out, harvestIfElse)

	closeTpl := spltoken.CloseAccountInstruction(tokenAccount, rentDestination, owner, accts.OwnerSigner)
	closeCpi, closeRem, err := patchedcpi.StaticCpi(closeTpl, nil)
	if err != nil {
		return nil, err
	}
	closeArgs, err := ifelse.Args(dust, ifelse.Cpi(closeCpi), ifelse.Skip)
	if err != nil {
		return nil, err
	}
	closeIfElse, err := s.IxIfElse(closeArgs, closeRem)
	if err != nil {
		return nil, err
	}
	out = append(out, closeIfElse)

	return out, nil
}

File: rust-sdk/tests/common/planners/dust_destroy.rs
//! Token-2022 dust destroy — mirrors [`go-sdk/examples/dust_destroy.go`](../../../../go-sdk/examples/dust_destroy.go).

use ifx_sdk::expr;
use ifx_sdk::if_else::{args, cpi, skip};
use ifx_sdk::patched_cpi::{
    build_raw_cpi, build_static_cpi, raw_cpi_patch, with_owner_signer,
};
use ifx_sdk::scratch::FrameScratch;
use ifx_sdk::ScratchError;
use solana_sdk::instruction::Instruction;
use solana_sdk::pubkey::Pubkey;

use crate::common::token2022::{
    burn_checked_instruction, close_account_instruction, harvest_withheld_instruction,
};

/// Raw balance cutoff (NOT UI amount).
pub const DUST_THRESHOLD_RAW: u64 = 1000;

#[derive(Clone, Debug)]
pub struct DustDestroyAccounts {
    pub mint: Pubkey,
    pub token_account: Pubkey,
    pub owner: Pubkey,
    pub owner_signer: bool,
    pub rent_destination: Pubkey,
}

pub fn plan_dust_destroy_instructions(
    scratch: &mut FrameScratch,
    accts: &DustDestroyAccounts,
) -> Result<Vec<Instruction>, ScratchError> {
    let DustDestroyAccounts {
        mint,
        token_account,
        owner,
        owner_signer,
        rent_destination,
    } = *accts;

    let mut out = Vec::new();
    out.push(scratch.ix_reset());

    let mut b = scratch.let_builder();
    let amount = b.spl_token2022_amount(token_account)?;
    let withheld = b.spl_token2022_transfer_fee_withheld(token_account)?;
    let decimals = b.spl_token2022_mint_decimals(mint)?;
    out.push(b.build_ix()?);

    let dust = expr::lt(expr::r(&amount), expr::u64(DUST_THRESHOLD_RAW));

    let burn_tpl = burn_checked_instruction(token_account, mint, owner, owner_signer);
    let burn_built = build_raw_cpi(
        &burn_tpl,
        &[
            raw_cpi_patch(1, &amount),
            raw_cpi_patch(9, &decimals),
        ],
    )?;
    out.push(scratch.ix_if_else(
        &args(
            expr::and(dust.clone(), expr::non_zero(expr::r(&amount))),
            cpi(burn_built.cpi.clone()),
            skip(),
        ),
        &burn_built.remaining,
    )?);

    let harvest_tpl = harvest_withheld_instruction(mint, token_account);
    let harvest_built = build_static_cpi(&harvest_tpl)?;
    out.push(scratch.ix_if_else(
        &args(
            expr::and(dust.clone(), expr::non_zero(expr::r(&withheld))),
            cpi(harvest_built.cpi.clone()),
            skip(),
        ),
        &harvest_built.remaining,
    )?);

    let close_tpl = close_account_instruction(token_account, rent_destination, owner, owner_signer);
    let close_ix = with_owner_signer(&close_tpl, owner, owner_signer);
    let close_built = build_static_cpi(&close_ix)?;
    out.push(scratch.ix_if_else(
        &args(dust, cpi(close_built.cpi.clone()), skip()),
        &close_built.remaining,
    )?);

    Ok(out)
}
